rotkehlchen/chain/bitcoin/__init__.py

```diff
@@ -9,57 +9,91 @@
+    yield _query_mempool_space
+
+
+def get_bitcoin_addresses_balances(
+        accounts: List[BTCAddress],
+) -> Dict[BTCAddress, FVal]:
+    """Queries bitcoin balance APIs for the balances of accounts
 
     May raise:
-    - RemotError if there is a problem querying blockchain.info or blockstream
+    - RemoteError if there is a problem querying blockchain.info or blockstream
     """
-    source = 'blockchain.info'
-    balances: Dict[BTCAddress, FVal] = {}
-    try:
-        if _have_bc1_accounts(accounts):
-            # if 1 account is bech32 we have to query blockstream. blockchaininfo won't work
-            source = 'blockstream'
-            balances = {}
-            for account in accounts:
-                url = f'https://blockstream.info/api/address/{account}'
-                response_data = request_get_dict(url=url, handle_429=True, backoff_in_seconds=4)
-                stats = response_data['chain_stats']
-                balance = int(stats['funded_txo_sum']) - int(stats['spent_txo_sum'])
-                balances[account] = satoshis_to_btc(balance)
-        else:
-            # split the list of accounts into sublists of 80 addresses per list to overcome:
-            # https://github.com/rotki/rotki/issues/3037
-            accounts_chunks = [accounts[x:x + 80] for x in range(0, len(accounts), 80)]
-            for accounts_chunk in accounts_chunks:
-                params = '|'.join(accounts_chunk)
-                btc_resp = request_get_dict(
-                    url=f'https://blockchain.info/multiaddr?active={params}',
-                    handle_429=True,
-                    # If we get a 429 then their docs suggest 10 seconds
-                    # https://blockchain.info/q
-                    backoff_in_seconds=10,
-                )
-                for entry in btc_resp['addresses']:
-                    balances[entry['address']] = satoshis_to_btc(FVal(entry['final_balance']))
-    except (
+    error: Union[Exception, None] = None
+    for api in _get_blockchain_api():
+        try:
+            balances = api(accounts)
+        except (
             requests.exceptions.RequestException,
             UnableToDecryptRemoteData,
             requests.exceptions.Timeout,
-    ) as e:
-        raise RemoteError(f'bitcoin external API request for balances failed due to {str(e)}') from e  # noqa: E501
-    except KeyError as e:
-        raise RemoteError(
-            f'Malformed response when querying bitcoin blockchain via {source}.'
-            f'Did not find key {e}',
-        ) from e
+            BlockchainNotSupported,
+        ) as e:
+            error = e
+            continue
+        except KeyError as e:
+            # Hacky way to get name of the source
+            source_name = '.'.join(api.__name__.split('_')[1:])
+            raise RemoteError(
+                f'Malformed response when querying bitcoin blockchain via {source_name}.'
+                f'Did not find key {e}',
+            ) from e
```

Comment:
I think KeyError is supposed to be raised here, as it's not really related to the failure of service we're querying.

Reply:
Hm, yes, probably you are right. Just pls verify that in case if an error like the issue describes appears, it is handled properly and execution continues